src/autodiscovery.js

```diff
@@ -0,0 +1,379 @@
+            const failingClientConfig = {
+                "m.homeserver": {
+                    state: AutoDiscovery.FAIL_ERROR,
+                    error: "Invalid identity server discovery response",
+
+                    // We'll provide the base_url that was previously valid for
+                    // debugging purposes.
+                    base_url: clientConfig["m.homeserver"].base_url,
+                },
+                "m.identity_server": {
+                    state: AutoDiscovery.FAIL_ERROR,
+                    error: "Invalid identity server discovery response",
+                    base_url: null,
+                },
+            };
+
+            // Step 5a: Make sure the URL is valid *looking*. We'll make sure it
+            // points to an identity server in Step 5b.
+            isUrl = this._sanitizeWellKnownUrl(
+                wellknown.raw["m.identity_server"]["base_url"],
+            );
+            if (!isUrl) {
+                logger.error("Invalid base_url for m.identity_server");
+                return Promise.resolve(failingClientConfig);
+            }
+
+            // Step 5b: Verify there is an identity server listening on the provided
+            // URL.
+            const isResponse = await this._fetchWellKnownObject(
+                `${isUrl}/_matrix/identity/api/v1`,
+            );
+            if (!isResponse || !isResponse.raw || isResponse.action !== "SUCCESS") {
+                logger.error("Invalid /api/v1 response");
+                return Promise.resolve(failingClientConfig);
+            }
+        }
+
+        // Step 6: Now that the identity server is valid, or never existed,
+        // populate the IS section.
+        if (isUrl && isUrl.length > 0) {
+            clientConfig["m.identity_server"] = {
+                state: AutoDiscovery.SUCCESS,
+                error: null,
+                base_url: isUrl,
+            };
+        }
+
+        // Step 7: Copy any other keys directly into the clientConfig. This is for
+        // things like custom configuration of services.
+        Object.keys(wellknown.raw)
+            .filter((k) => k !== "m.homeserver" && k !== "m.identity_server")
+            .map((k) => clientConfig[k] = wellknown.raw[k]);
+
+        // Step 8: Give the config to the caller (finally)
+        return Promise.resolve(clientConfig);
+    }
+
+    /**
+     * Sanitizes a given URL to ensure it is either an HTTP or HTTP URL and
+     * is suitable for the requirements laid out by .well-known auto discovery.
+     * If valid, the URL will also be stripped of any trailing slashes.
+     * @param {string} url The potentially invalid URL to sanitize.
+     * @return {string|boolean} The sanitized URL or a falsey value if the URL is invalid.
+     * @private
+     */
+    static _sanitizeWellKnownUrl(url) {
+        if (!url) return false;
+
+        try {
+            const parsed = parseURL(url);
```

Comment:
I believe that whatwg url parsing is supported in all browsers we support, so you could just use the [`URL` constructor](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL) here, [like so](https://url.spec.whatwg.org/#example-5434421b). That might be what the whatwg-url dependency will fall back to, but it still pulls in the dependency...

Reply:
Although not a solution, it would also allow us to ignore the export ES6 vs ES5/uglify-es problem for a bit longer.

Reply:
Node doesn't support the URL constructor though, which is why the dependency is pulled in. I'd love to use the URL constructor, but people do use the js-sdk outside of a web browser.

Reply:
Seems supported since 6.x in node?
https://nodejs.org/docs/latest-v6.x/api/url.html#url_the_whatwg_url_api

Reply:
I hate it when the internet lies to me :/

I'll give it a go